core/lib/zksync_core/src/sync_layer/batch_status_updater/mod.rs

```diff
@@ -0,0 +1,451 @@
+        AggregatedActionType::Execute => "executed",
+    }
+}
+
+/// Represents a change in the batch status.
+/// It may be a batch being committed, proven or executed.
+#[derive(Debug)]
+struct BatchStatusChange {
+    number: L1BatchNumber,
+    l1_tx_hash: H256,
+    happened_at: DateTime<Utc>,
+}
+
+#[derive(Debug, Default)]
+struct StatusChanges {
+    commit: Vec<BatchStatusChange>,
+    prove: Vec<BatchStatusChange>,
+    execute: Vec<BatchStatusChange>,
+}
+
+impl StatusChanges {
+    /// Returns true if there are no status changes.
+    fn is_empty(&self) -> bool {
+        self.commit.is_empty() && self.prove.is_empty() && self.execute.is_empty()
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum UpdaterError {
+    #[error("JSON-RPC error communicating with main node")]
+    Web3(#[from] ClientError),
+    #[error("Internal error")]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<zksync_dal::SqlxError> for UpdaterError {
+    fn from(err: zksync_dal::SqlxError) -> Self {
+        Self::Internal(err.into())
+    }
+}
+
+#[async_trait]
+trait MainNodeClient: fmt::Debug + Send + Sync {
+    /// Returns any miniblock in the specified L1 batch.
+    async fn resolve_l1_batch_to_miniblock(
+        &self,
+        number: L1BatchNumber,
+    ) -> Result<Option<MiniblockNumber>, ClientError>;
+
+    async fn block_details(
+        &self,
+        number: MiniblockNumber,
+    ) -> Result<Option<api::BlockDetails>, ClientError>;
+}
+
+#[async_trait]
+impl MainNodeClient for HttpClient {
+    async fn resolve_l1_batch_to_miniblock(
+        &self,
+        number: L1BatchNumber,
+    ) -> Result<Option<MiniblockNumber>, ClientError> {
+        let request_latency = FETCHER_METRICS.requests[&FetchStage::GetMiniblockRange].start();
+        let number = self
+            .get_miniblock_range(number)
+            .await?
+            .map(|(start, _)| MiniblockNumber(start.as_u32()));
+        request_latency.observe();
+        Ok(number)
+    }
+
+    async fn block_details(
+        &self,
+        number: MiniblockNumber,
+    ) -> Result<Option<api::BlockDetails>, ClientError> {
+        let request_latency = FETCHER_METRICS.requests[&FetchStage::GetBlockDetails].start();
+        let details = self.get_block_details(number).await?;
+        request_latency.observe();
+        Ok(details)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+struct UpdaterCursor {
+    last_executed_l1_batch: L1BatchNumber,
+    last_proven_l1_batch: L1BatchNumber,
+    last_committed_l1_batch: L1BatchNumber,
+}
+
+impl UpdaterCursor {
+    async fn new(storage: &mut StorageProcessor<'_>) -> anyhow::Result<Self> {
+        let first_l1_batch_number = projected_first_l1_batch(storage).await?;
+        // Use the snapshot L1 batch, or the genesis batch if we are not using a snapshot.
+        let starting_l1_batch_number = L1BatchNumber(first_l1_batch_number.saturating_sub(1));
+
+        let last_executed_l1_batch = storage
```

Comment:
This is a bit confusing, The last L1 batch from a snapshot is not necessarily proven yet, thought I don't yet see that this breaks anything.

Reply:
Added a comment why this is OK.